### PR TITLE
Adding NMF as auxiliary loss 

### PR DESCRIPTION
**Pull request by mohazahran**
@arvindsrikantan 
This PR is still rough with a lot of hard coding. But in principle it should be working. An early review would help if you have time.

Update:
The PR adds the following:
- Adding auxiliary loss ability to ml4ir to be used for NMF
- Adding basic cross entropy loss which is useful for as NMF auxiliary loss.

### CODE REVIEW

**Review by arvindsrikantan — COMMENTED** [review]
Thanks for sharing early @mohazahran . Let's create subclasses for single vs multiple losses where appropriate so that we don't enforce multiple labels always

File: python/ml4ir/applications/ranking/model/losses/listwise_losses.py [hunk 1]
```diff
@@ -32,7 +35,104 @@ def _loss_fn(y_true, y_pred):
...
+            """
+            Shapes
+            ------
+            y_true : [batch_size, num_classes]
+            y_pred : [batch_size, num_classes]
+            mask : [batch_size, num_classes]
+            """
+
+            if is_aux_loss:
+                y_true_softmax = tf.math.softmax(y_true)  #convert to a prod. dist
```
Comment (arvindsrikantan) [inline python/ml4ir/applications/ranking/model/losses/listwise_losses.py:125]: Typo: `prob.`. I prefer the dull word though

File: python/ml4ir/base/model/relevance_model.py [hunk 2]
```diff
@@ -48,7 +48,11 @@ def __init__(
         freeze_layers_list: list = [],
         compile_keras_model: bool = False,
         output_name: str = "score",
+        aux_output_name: str = "aux_score",
```
Comment (arvindsrikantan) [inline python/ml4ir/base/model/relevance_model.py:51]: I feel we could've created a subclass for `RelevanceModelAuxLoss`. Bad name suggestion on my part, but something along those lines

File: python/ml4ir/base/model/relevance_model.py [hunk 3]
```diff
@@ -431,7 +469,8 @@ def fit(
             where key is metric name and value is floating point metric value.
             This dictionary will be used for experiment tracking for each ml4ir run
         """
-
+        if self.feature_config.aux_label: #val_ranking_score_new_MRR
+            monitor_metric = 'val_' + self.output_name + '_' + monitor_metric
```
Comment (arvindsrikantan) [inline python/ml4ir/base/model/relevance_model.py:473]: Why is this needed?

**Review by arvindsrikantan — COMMENTED** [review]
Added some initial comments

File: python/ml4ir/applications/ranking/pipeline.py [hunk 4]
```diff
@@ -90,13 +92,20 @@ def get_relevance_model(self, feature_layer_keys_to_fns={}) -> RelevanceModel:
...
+            losses = {self.args.output_name: loss, self.args.aux_output_name:aux_loss}
+
 
         # Define scorer
         scorer: ScorerBase = RelevanceScorer(
             feature_config=self.feature_config,
             model_config=self.model_config,
             interaction_model=interaction_model,
-            loss=loss,
+            loss=losses,
```
Comment (arvindsrikantan) [inline python/ml4ir/applications/ranking/pipeline.py:108]: You need to change the `__init__` method signature for RelevanceScorer

File: python/ml4ir/applications/ranking/model/losses/listwise_losses.py [hunk 5]
```diff
@@ -32,7 +35,104 @@ def _loss_fn(y_true, y_pred):
...
+        is_aux_loss = False
+        if kwargs.get("is_aux_loss"):
+            is_aux_loss = True
+        batch_size = kwargs.get("batch_size", 1)
+
+        def _loss_fn(y_true, y_pred):
+            """
+            Shapes
+            ------
+            y_true : [batch_size, num_classes]
```
[the remark this answers is not in the document]
Comment (mohazahran) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:119]: This loss expect y-true to be a prob dist. I explicitly apply a softmax for aux loss. I can enforce it for non aux loss as well

File: python/ml4ir/base/model/relevance_model.py [hunk 6]
```diff
@@ -120,29 +130,57 @@ def __init__(
...
 
-            # Create model with functional Keras API
-            self.model = Model(inputs=inputs, outputs={self.output_name: scores})
-            self.model.output_names = [self.output_name]
-
-            # Get loss fn
-            loss_fn = scorer.loss.get_loss_fn(**metadata_features)
+            if self.feature_config.get_aux_label():
+                # Create model with functional Keras API
+                self.model = Model(inputs=inputs, outputs={self.output_name: scores, self.output_name+"_aux": scores})
```
[the remark this answers is not in the document]
Comment (mohazahran) [reply python/ml4ir/base/model/relevance_model.py:135]: of course. I changed it but did not update the PR. Thanks for the callout.

File: python/ml4ir/base/model/relevance_model.py (hunk 3 above)
Comment (mohazahran) [reply python/ml4ir/base/model/relevance_model.py:473]: We need to specify the metric to be tracked in the validation set.

File: python/ml4ir/applications/ranking/tests/test_losses.py [hunk 7]
```diff
@@ -21,6 +21,10 @@ def setUp(self):
             [[0., 0., 1., 0., 0.],
```
Comment (arvindsrikantan) [inline python/ml4ir/applications/ranking/tests/test_losses.py:21]: @mohazahran , can you add a serving end to end test too?

File: python/ml4ir/applications/ranking/model/losses/listwise_losses.py [hunk 8]
```diff
@@ -32,7 +35,104 @@ def _loss_fn(y_true, y_pred):
...
+            if is_aux_loss:
+                y_true_softmax = tf.math.softmax(y_true)  #convert to a prod. dist
+                # masking zeros for the log op
+                zero = tf.constant(0, dtype=tf.float32)
+                non_zero = tf.not_equal(y_pred, zero)
+                # remove all the zero entries from the y_pred (corresponds to padded records)
+                y_pred_non_zero = tf.boolean_mask(y_pred, non_zero)
+                # retain values in y_true corresponding to non zero values in y_pred
+                y_true_softmax_masked = tf.boolean_mask(y_true_softmax, non_zero)
+                return tf.math.divide(-tf.reduce_sum(y_true_softmax_masked * tf.math.log(y_pred_non_zero)), tf.constant(batch_size, dtype=tf.float32))
```
[the remark this answers is not in the document]
Comment (mohazahran) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:133]: I'm manually normalizing by the batch_size since we're computing the cross entropy without Keras wrappers which usually does this normalization.

[the remark this answers is not in the document]
Comment (arvindsrikantan) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:133]: I don't think loss is normalized by number of samples. It's just sum over all samples

File: python/ml4ir/applications/ranking/model/losses/listwise_losses.py [hunk 9]
```diff
@@ -32,7 +35,104 @@ def _loss_fn(y_true, y_pred):
             y_pred : [batch_size, num_classes]
             mask : [batch_size, num_classes]
             """
-            return cce(y_true, tf.math.multiply(y_pred, mask))
+
+            #Fixme
+            """
+            Queries with ties in the highest scores would have multiple one's in the 1-hot vector.
```
Comment (arvindsrikantan) [inline python/ml4ir/applications/ranking/model/losses/listwise_losses.py:41]: Should we just scale it by the number of docs with highest score?

File: python/ml4ir/applications/ranking/model/losses/listwise_losses.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (arvindsrikantan) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:119]: You're converting it to a probability distribution, so you don't expect one at the input right?

File: python/ml4ir/base/model/relevance_model.py [hunk 10]
```diff
@@ -120,29 +130,57 @@ def __init__(
...
+
+                lossWeights = {self.output_name: primary_loss_weight, self.aux_output_name: aux_loss_weight}
+
+                # Get metric objects
+                metrics_impl: List[Union[str, kmetrics.Metric]] = get_metrics_impl(
+                    metrics=metrics, feature_config=feature_config, metadata_features=metadata_features
+                )
+                metrics_impl_aux: List[Union[str, kmetrics.Metric]] = get_metrics_impl(
+                    metrics=metrics, feature_config=feature_config, metadata_features=metadata_features
+                )
```
Comment (arvindsrikantan) [inline python/ml4ir/base/model/relevance_model.py:154]: Do we need this?

File: python/ml4ir/base/model/scoring/scoring_model.py [hunk 11]
```diff
@@ -227,6 +227,10 @@ def final_activation_op(self, scores, metadata_features):
             Tensor object produced by applying the final activation function
             to the scores computed by the model
         """
+        if isinstance(self.loss, dict):
```
Comment (arvindsrikantan) [inline python/ml4ir/base/model/scoring/scoring_model.py:230]: Why not enforce loss to be a dict? Will it need a lot of refactoring?

File: python/ml4ir/applications/ranking/model/losses/listwise_losses.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mohazahran) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:119]: Both, y_pred and y_true should be prob dist. However, since the ranking pipeline already converts the predictions to a prob. dist before computing the loss, I'm explicitly just converting y_true to a prob dist.

File: python/ml4ir/base/model/relevance_model.py (hunk 2 above)
Comment (mohazahran) [reply python/ml4ir/base/model/relevance_model.py:51]: I believe that would be a nicer design going forward if we need to have N objectives. I believe now this might complicate things at the moment.

File: python/ml4ir/base/model/scoring/scoring_model.py (hunk 11 above)
Comment (mohazahran) [reply python/ml4ir/base/model/scoring/scoring_model.py:230]: I think it's doable, I just didn't feel the need to make that change at the moment.

File: python/ml4ir/applications/ranking/model/losses/listwise_losses.py (hunk 9 above)
Comment (mohazahran) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:41]: Why do you think that would be useful?

File: python/ml4ir/applications/ranking/model/losses/listwise_losses.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (arvindsrikantan) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:119]: Yes, I understand. I wanted you to add that to the documentation :)

File: python/ml4ir/applications/ranking/model/losses/listwise_losses.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (arvindsrikantan) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:133]: Ah my bad. I was assuming the learning rate folds in the normalization. That doesn't seem to be the case. I'm marking this as resolved. Thanks for the discussion

**Review by arvindsrikantan — COMMENTED** [review]
Want to hear your thoughts on scenarios where there are ties in the aux label. Looks good overall

File: python/ml4ir/applications/ranking/model/losses/listwise_losses.py (hunk 9 above)
Comment (arvindsrikantan) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:41]: @mohazahran , ping on this

Comment (mohazahran) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:41]: I don't quite get your point. Scaling by the number of ties, that would make the loss bigger for cases with many ties, right? How's that would help.

Comment (arvindsrikantan) [reply python/ml4ir/applications/ranking/model/losses/listwise_losses.py:41]: Ah, I see the ambiguity. I meant to scale down (1/num_ties )* loss

File: python/ml4ir/applications/ranking/tests/test_losses.py [hunk 12]
```diff
@@ -21,6 +21,10 @@ def setUp(self):
...
              [0., 1., 0., 0., 0.],
              [0., 1., 0., 0., 0.]])
+        self.y_true_aux = tf.constant(
+            [[0.34, 0.81, 0.22, -0.05, -0.67],
+             [-1.35, -0.75, -0.37, 1.5467, 0.8],
+             [-1.4356, 0.75, -0.6857, -0.689, 0.089]])
```
Comment (arvindsrikantan) [inline python/ml4ir/applications/ranking/tests/test_losses.py:27]: @mohazahran , last request. Can you add a test case with ties? Everything else is good to go 🚀
